lib/ptracer.h

```diff
@@ -0,0 +1,471 @@
+
+/* ---------------- DATA STRUCTURES ---------------- */
+
+/* What type of frame we are dealing with? Is it just a normal
+   C function or Lua C Function? */
+typedef enum frame_type {
+    PALLENE_TRACER_FRAME_TYPE_C,
+    PALLENE_TRACER_FRAME_TYPE_LUA
+} frame_type_t;
+
+/* Details of the callee function (name, where it is from etc.) */
+/* Optimization Tip: Try making it 'static' for every funciton. */
+typedef struct pt_fn_details {
+    const char *const fn_name;
+    const char *const filename;
+} pt_fn_details_t;
+
+/* A single frame representation. */
+typedef struct pt_frame {
+    frame_type_t type;
+    int line;
+
+    union {
+        pt_fn_details_t *details;
+        lua_CFunction frame_sig;
+    } shared;
+} pt_frame_t;
+
+/* Out stack is fully heap-allocated stack. We need some structure to hold
+   the stack information. This structure will be an Userdatum. */
+/* 'cont' stands for 'container'. */
+typedef struct pt_cont {
+    pt_frame_t *stack;
+    int count;
+} pt_cont_t;
+
+/* ---------------- DATA STRUCTURES END ---------------- */
+
+/* ---------------- DECLARATIONS ---------------- */
+
+#ifdef __cplusplus
+extern "C" {
+#endif // __cplusplus
+
+/* Pushes a frame to the stack. The frame structure is self-managed for every function. */
+void pallene_tracer_frameenter(lua_State *L, pt_cont_t *cont, pt_frame_t *restrict frame);
+
+/* Set's line number to the topmost frame in the stack. */
+void pallene_tracer_setline(pt_cont_t *cont, int line);
+
+/* Removes the last frame from the stack. */
+void pallene_tracer_frameexit(pt_cont_t *cont);
+
+/* Pallene Tracer explicit traceback function to show Pallene call-stack
+   tracebacks. */
+int  pallene_tracer_debug_traceback(lua_State *L);
+
+/* When we encounter an error, `pallene_tracer_frameexit()` may not
+   get called. Therefore, the stack will get corrupted if the previous
+   call-frames are not removed. The finalizer function makes sure that
+   it does not happen. It's guardian angel. */
+/* The finalizer function will be called from a to-be-closed value (since
+   Lua 5.4). If you are using Lua version prior 5.4, you are outta luck. */
+int pallene_tracer_finalizer(lua_State *L);
+
+/* Initializes the Pallene Tracer. The initialization refers to creating the stack
+   if not created, preparing the traceback fn and finalizers. */
+/* NOTE: Pushes the finalizer object to the stack. The object has to be closed
+   everytime you are in a Lua C funciton using `lua_toclose(L, idx)` function. */
+pt_cont_t *pallene_tracer_init(lua_State *L);
+
+/* When Pallene call-stack is overflowed. */
+l_noret pallene_tracer_runtime_callstack_overflow_error(lua_State *L);
+
+#ifdef __cplusplus
+}
+#endif // __cplusplus
+
+/* ---------------- DECLARATIONS END ---------------- */
+
+#endif // PALLENE_TRACER_H
+
+#if defined(PT_IMPLEMENTATION) && !defined(PT_IMPLEMENTED)
+#define PT_IMPLEMENTED
+
+/* ---------------- PRIVATE ---------------- */
```

Comment:
I'm worried that these private functions live in a header file that we will include in another C file. If we are going with this STB header approach, shouldn't they have names that are less likely to clash with functions created by the user?

Reply:
Working on it.